packages/cluster-ui/src/transactionsPage/utils.ts

```diff
@@ -69,38 +81,94 @@ export const filterTransactions = (
+  nodeRegions: { [key: string]: string },
 ): { transactions: Transaction[]; activeFilters: number } => {
   if (!filters)
     return {
       transactions: data,
       activeFilters: 0,
     };
   const timeValue = getTimeValueInSeconds(filters);
-  const filtersStatus = [
-    timeValue && timeValue !== "empty",
-    filters.app !== "All",
-  ];
-  const activeFilters = filtersStatus.filter(f => f).length;
-
-  const filteredTransactions = data.filter((t: Transaction) => {
-    const matchAppNameExactly = t.stats_data.app === filters.app;
-    const filterIsSetToAll = filters.app === "All";
-    const filterIsInternalMatchByPrefix =
-      filters.app === internalAppNamePrefix &&
-      t.stats_data.app.includes(filters.app);
-    const validateTransaction = [
-      matchAppNameExactly || filterIsSetToAll || filterIsInternalMatchByPrefix,
-      t.stats_data.stats.service_lat.mean >= timeValue || timeValue === "empty",
-    ];
-    return validateTransaction.every(f => f);
-  });
+  const regions = filters.regions.length > 0 ? filters.regions.split(",") : [];
+  const nodes = filters.nodes.length > 0 ? filters.nodes.split(",") : [];
+
+  const activeFilters = calculateActiveFilters(filters);
+  const filteredTransactions = data
+    .filter(
+      (t: Transaction) =>
+        filters.app === "All" ||
+        t.stats_data.app === filters.app ||
+        (filters.app === internalAppNamePrefix &&
+          t.stats_data.app.includes(filters.app)),
+    )
+    .filter(
+      (t: Transaction) =>
+        t.stats_data.stats.service_lat.mean >= timeValue ||
+        timeValue === "empty",
+    )
+    .filter((t: Transaction) => {
+      if (regions.length == 0 && nodes.length == 0) return true;
+      let foundRegion: boolean = regions.length == 0;
+      let foundNode: boolean = nodes.length == 0;
+
+      getStatementsById(t.stats_data.statement_ids, statements).some(stmt => {
+        stmt.stats.nodes.some(node => {
+          if (foundRegion || regions.includes(nodeRegions[node.toString()])) {
+            foundRegion = true;
+          }
+          if (foundNode || nodes.includes("n" + node)) {
+            foundNode = true;
+          }
+          if (foundNode && foundRegion) return true;
+        });
+      });
+
+      return foundRegion && foundNode;
+    });
 
   return {
     transactions: filteredTransactions,
     activeFilters,
   };
 };
 
+// For each transaction, generate the list of regions and nodes all
+// its statements were executed on.
+// E.g. of one element of the list: `gcp-us-east1 (n1, n2, n3)`
+export const generateRegionNode = (
+  transaction: Transaction,
+  statements: Statement[],
+  nodeRegions: { [p: string]: string },
+): string[] => {
+  const regions: { [region: string]: Set<number> } = {};
+  getStatementsById(transaction.stats_data.statement_ids, statements).forEach(
+    stmt => {
+      stmt.stats.nodes.forEach(node => {
+        if (Object.keys(regions).includes(nodeRegions[node.toString()])) {
```

Comment:
nit: would be nice to extract `nodeRegions[node.toString()]` to its own variable since it's used multiple times in this func.

Reply:
done